### ISSUE TITLE
Design Ideas

### ISSUE THREAD

**Issue by influx6** [murlokswarm/linux#1]
Since am majorly on linux. This is definitely something that interests me in working.
Thinking on the major step to get this working is as I think, is to use Webkit in GTK.

Luckily we already have go bindings for using the webkit API and I think that will help alot in getting the 
work we need to do less intense.

Webkit2GTK API:
https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebView.html

Go webkit bindings:
https://github.com/mattn/go-webkit
https://github.com/sourcegraph/go-webkit2

Go GTK bindings:
https://github.com/mattn/go-gtk

My plan is to study the binding code and the mac driver code for murlock and see what I can come up with. If we can get a rendering and events connected then that would be a awesome milestone.

**Comment by maxence-charriere** [comment]
That would be awesome,
Let me know if you need something, I'll try to answer you the best I can.
👍

**Comment by influx6** [comment]
Will do.

**Comment by influx6** [comment]
Quick question: @maxence-charriere 

How do you get the content of your templates into the mac drivers view and update them? Any major code. I observer you have some javascript functions which you introduce into a new driver. Does that provide the means to interoperate between the go and javascript side?

**Comment by maxence-charriere** [comment]
Must implement the interface [app.Contexter](https://godoc.org/github.com/murlokswarm/app#Contexter) for your window.

here is a pseudo instruction on how it is done depending of the action

## window.Mount()
1. Get the HTML from a component:
[markup.Markup(*component*)](https://godoc.org/github.com/murlokswarm/markup#Markup) 
2. Use the web view API to call javascript [function Mount(id, markup)](https://github.com/murlokswarm/app/blob/master/js.go#L15)
e.g. https://github.com/murlokswarm/mac/blob/master/window.go#L104

## window.Render()
For render it is a little more complicated.
Render give you a list of markup.Sync which describe a list of node to sync.
1. Iterate on syncs
2. if a sync describe a [full sync](https://github.com/murlokswarm/mac/blob/master/window.go#L124):
- get the HTML ```s.Node.Markup()
- Use the web view API to call javascript [function RenderFull(id, markup)](https://github.com/murlokswarm/mac/blob/master/window.go#L133)
2. (bis) if sync requires just the attributes to be modified:
- translate sync.Attributes [to JSON](https://github.com/murlokswarm/mac/blob/master/window.go#L141)
- Use the web view API to call javascript [function RenderAttributes(id, attrs)](https://github.com/murlokswarm/mac/blob/master/window.go#L146)

**Comment by influx6** [comment]
Awesome, thanks man.

**Comment by maxence-charriere** [comment]
i updated things
also you will have to init the web view with the basic html.
take a look on https://github.com/murlokswarm/mac/blob/master/window.go#L44

**Comment by influx6** [comment]
Alright, thanks man. Very informative.

**Comment by influx6** [comment]
His windows working for you now?

**Comment by maxence-charriere** [comment]
Not yet but I'm on a good way. I solved bidirectional communication between a dll and go. And was able to compile the c++ code to communicate with a UWP.
Now on the packaging to see if that rally works.

**Comment by influx6** [comment]
Awesome bro. Keep it rocking.

**Comment by influx6** [comment]
How does your current EdgeHTML and macdriver call into Go code @maxence-charriere ?

**Comment by maxence-charriere** [comment]
On mac it uses CGo export func.
See https://github.com/murlokswarm/mac/blob/master/driver.go#L112

**Comment by influx6** [comment]
Ah! I see. Will look more into it. Thanks bro.

Mr Alex.

Software Engineer Consultant,
Ardan Studios.
www.ardanstudios.com

Chief Information Officer (CIO)
Dream Bench Technologies
www.dreambench.io

On Mon, Feb 13, 2017 at 8:30 PM, Maxence Charriere <notifications@github.com

**Comment by cryptix** [comment]
@influx6 any way I can help you with this? I'd love to test and do code review.

**Comment by influx6** [comment]
Hey, thanks man. Being locked down due to some personal issues have not had
time to work on this yet. Currently code has not reach any major point.

Will let you know asap once i have something work.

Also am thinking of looking into QT as a better alternative due to its wide
reach.

Luckily we also have a qt binding https://github.com/therecipe/qt

**Comment by mubaris** [comment]
@influx6 @maxence-charriere I would like to work with this project.

**Comment by influx6** [comment]
Please bro. Take it on.

I was not really successfully with the QT binding and work is taking alot
of my time. So hopefully once am free I can look at this again.

If you can work on a gtk webview that matches the Driver for the
murlockswarm then that wpuld be awesome.

**Comment by cryptix** [comment]
@mubaris: still down for code review but not enough spare time to take the lead right now.

**Comment by mubaris** [comment]
@cryptix @influx6 @maxence-charriere I can start contributing. But the thing is that I am not sure how to start and where to start. If you guys can just guide me about beginning of this project, I can start working on this. Then I will be good to go. A detailed guide would be good. Is it possible that after I read the codes of [murlokswarm/app](http://github.com/murlokswarm/app) and [murlokswarm/mac](github.com/murlokswarm/mac), I will be able to do it by myself?

**Comment by cryptix** [comment]
Yup. afaict the trick to make this work is the [Driver Interface](https://godoc.org/github.com/murlokswarm/app#Driver), that's what https://github.com/murlokswarm/mac implements for mac (duh). If you can take that as inspiration/input to rig up a webkit2gtk version you should be good to go.

**Comment by influx6** [comment]
Exactly as @Henry points out. I have documented the different gtk go
libraries and you can checkout sourcegraph as well as they have a webview
sample you can look at. All the best mate.

Best Regards,
Alex.

Chief Information Officer (CIO)
Dream Bench Technologies
www.dreambench.io (alex.ewetumo@dreambench.com)

**Comment by gauthamzz** [comment]
I would like to also help,if that is fine :smile:

**Comment by maxence-charriere** [comment]
Hey guys, I'm so happy to see that this keep your interest.
As @cryptix said, the key is to implement the [Driver](https://godoc.org/github.com/murlokswarm/app#Driver).

The [mac](https://github.com/murlokswarm/mac) package is a good start to understand how it works.

Maybe before implementing a driver we should proceed step by step for a proof of concept.
1. Display a window with go.
2. Embeds a web view (please a webkit or cef one) in this window that shows a website.
3. Display custom html in the web view from a Go call.
4 call a Go func from javascript from the web view.

Just focus by doing these steps and then we can shape it to make it fit into the Driver interface.

**Comment by mubaris** [comment]
Cool :sunglasses:. @maxence-charriere That's what I wanted. I will be using GTK with WebKit.

**Comment by maxence-charriere** [comment]
if you are really motivated, send me a PR for each steps. I ll review and merge at each one.
